File: server-app/services/topic_service.py
import logging
from repositories.topic_repo import TopicRepository
from models.topic import Topic
from models.schemas import TopicContent
from repositories.chapter_repo import ChapterRepository
from repositories.module_repo import ModuleRepository
from repositories.subject_repo import SubjectRepository
from repositories.course_repo import CourseRepository
from utils.gemini_helper import GeminiHelper

logger = logging.getLogger(__name__)

class TopicService:
    def __init__(self):
        self.topic_repo = TopicRepository()
        self.chapter_repo = ChapterRepository()
        self.module_repo = ModuleRepository()
        self.subject_repo = SubjectRepository()
        self.course_repo = CourseRepository()
        self.gemini_helper = GeminiHelper()

    def generate_topics(self, course_id, subject_id, module_id, chapter_id):
        try:
            logger.info(f"Starting topic generation for chapter_id: {chapter_id}")
            
            chapter = self.chapter_repo.get_chapters_by_module_id(module_id)
            if not chapter:
                logger.error(f"Chapter not found for module_id: {module_id}")
                raise Exception("Chapter not found")
            
            chapter_info = [item for item in chapter if item.id == chapter_id][0]
            logger.debug(f"Found chapter: {chapter_info.name}")

            module = self.module_repo.get_module_by_id(module_id)
            if not module:
                logger.error(f"Module not found with id: {module_id}")
                raise Exception("Module not found")
            
            subject = self.subject_repo.get_subjects_by_course_id(course_id)
            if not subject:
                logger.error(f"Subject not found for course_id: {course_id}")
                raise Exception("Subject not Found")
            subject_name = [item for item in subject if item.id == subject_id][0].name

            prompt = f"""Generate a detailed list of topics for the chapter '{chapter_info.name}' 
            in module '{module.name}' under subject '{subject_name}'.
            Requirements:
            1. Each topic should be a specific learning point
            2. Topics should progress logically from basic to advanced
            3. Include 5-10 topics per chapter
            4. Include relevant mind maps where appropriate
            5. Topics should be more specific than chapter content
            6. Use clear, concise academic terminology
            7. Ensure comprehensive coverage of the chapter
            """
            
            logger.debug("Calling Gemini API for content generation")
            response = self.gemini_helper.generate_content(
                prompt,
                response_schema=TopicContent
            )
            logger.debug(f"Received response from Gemini: {response}")
            
            topics_added = 0
            for topic_name in response['topics']:
                topic = Topic(
                    chapter_id=chapter_id,
                    name=topic_name
                )
                self.topic_repo.add_topic(topic)
                topics_added += 1
            
            logger.info(f"Successfully added {topics_added} topics")
            return {"message": f"Successfully generated {topics_added} topics"}
            
        except Exception as e:
            logger.error(f"Error in generate_topics: {str(e)}", exc_info=True)
            raise Exception(f"Error generating topics: {str(e)}")

    def get_topics_by_chapter_id(self, chapter_id):
        return [topic.to_dict() for topic in self.topic_repo.get_topics_by_chapter_id(chapter_id)]
    
    def get_topic_by_id(self, topic_id):
        topic = self.topic_repo.get_topic_by_id(topic_id)
        return topic.to_dict() if topic else None

File: go-server/services/topic_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"go-server/models"
	"go-server/repositories"
	"go-server/utils"

	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

type TopicService interface {
	GenerateTopics(courseID, subjectID, chapterID uint) ([]models.Topic, error)
	GetTopicsByChapterID(chapterID uint) ([]models.Topic, error)
	GetTopicByID(topicID uint) (*models.Topic, error)
	CreateTopics(topics []models.Topic) error
	UpdateTopic(topicID uint, name string) (*models.Topic, error)
	DeleteTopic(topicID uint) error
	DeleteTopicsByChapterID(chapterID uint) error
}

type TopicServiceImpl struct {
	topicRepo    repositories.TopicRepository
	chapterRepo  repositories.ChapterRepository
	subjectRepo  repositories.SubjectRepository
	courseRepo   repositories.CourseRepository
	geminiHelper *utils.GeminiHelper
}

func NewTopicService(topicRepo repositories.TopicRepository, chapterRepo repositories.ChapterRepository, subjectRepo repositories.SubjectRepository, courseRepo repositories.CourseRepository) TopicService {
	return &TopicServiceImpl{
		topicRepo:    topicRepo,
		chapterRepo:  chapterRepo,
		subjectRepo:  subjectRepo,
		courseRepo:   courseRepo,
		geminiHelper: utils.NewGeminiHelper(),
	}
}

// GenerateTopics generates topics for a chapter using Gemini AI
func (s *TopicServiceImpl) GenerateTopics(courseID, subjectID, chapterID uint) ([]models.Topic, error) {
	logrus.Infof("Starting topic generation for chapter_id: %d, subject_id: %d, course_id: %d", chapterID, subjectID, courseID)

	// Get chapter
	chapter := &models.Chapter{}
	if err := s.chapterRepo.GetByID(chapterID, chapter); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("chapter not found")
		}
		return nil, errors.New("failed to get chapter")
	}

	// Verify chapter belongs to subject
	if chapter.SubjectID != subjectID {
		return nil, errors.New("chapter does not belong to the specified subject")
	}

	// Get subject
	subject := &models.Subject{}
	if err := s.subjectRepo.GetByID(subjectID, subject); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("subject not found")
		}
		return nil, errors.New("failed to get subject")
	}

	// Verify subject belongs to course
	if subject.CourseID != courseID {
		return nil, errors.New("subject does not belong to the specified course")
	}

	// Get course
	course := &models.Course{}
	if err := s.courseRepo.GetByID(courseID, course); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("course not found")
		}
		return nil, errors.New("failed to get course")
	}

	logrus.Debugf("Found chapter: %s, subject: %s, course: %s", chapter.Name, subject.Name, course.Name)

	// Check if topics already exist
	existingTopics, err := s.topicRepo.GetTopicsByChapterID(chapterID)
	if err != nil {
		return nil, errors.New("failed to check existing topics")
	}

	// If topics exist, return them
	if len(existingTopics) > 0 {
		logrus.Infof("Topics already exist for chapter %s, returning existing topics", chapter.Name)
		return existingTopics, nil
	}

	// Generate topics using Gemini
	ctx := context.Background()
	prompt := fmt.Sprintf(`Generate a comprehensive list of topics for the chapter '%s' in subject '%s' for the course '%s'.
	Consider the following:
	1. Include topics from basic to advanced level within this chapter
	2. Each topic should be a distinct subtopic within the chapter
	3. Topics should follow a logical learning progression
	4. Generate maximum 6 topics for the chapter
	5. Keep topic names concise and clear
	6. Focus on specific concepts that can be taught individually`, chapter.Name, subject.Name, course.Name)

	topicNames, err := s.geminiHelper.GenerateTopics(ctx, chapter.Name, subject.Name, course.Name)
	if err != nil {
		logrus.Errorf("Error generating topics with Gemini: %v", err)
		return nil, errors.New("failed to generate topics")
	}

	logrus.Infof("Generated %d topics for chapter %s", len(topicNames), chapter.Name)

	// Create topic models
	var topics []models.Topic
	for _, name := range topicNames {
		topics = append(topics, models.Topic{
			Name:      name,
			ChapterID: chapterID,
		})
	}

	// Save topics
	if err := s.topicRepo.CreateTopics(topics); err != nil {
		logrus.Errorf("Error saving topics: %v", err)
		return nil, errors.New("failed to save topics")
	}

	logrus.Infof("Successfully created %d topics for chapter: %s", len(topics), chapter.Name)
	return topics, nil
}

// GetTopicsByChapterID returns all topics for a chapter
func (s *TopicServiceImpl) GetTopicsByChapterID(chapterID uint) ([]models.Topic, error) {
	topics, err := s.topicRepo.GetTopicsByChapterID(chapterID)
	if err != nil {
		logrus.Errorf("Error getting topics by chapter ID: %v", err)
		return nil, errors.New("failed to get topics")
	}
	return topics, nil
}

// GetTopicByID returns a specific topic
func (s *TopicServiceImpl) GetTopicByID(topicID uint) (*models.Topic, error) {
	topic := &models.Topic{}
	if err := s.topicRepo.GetByID(topicID, topic); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("topic not found")
		}
		logrus.Errorf("Error getting topic: %v", err)
		return nil, errors.New("failed to get topic")
	}
	return topic, nil
}

// CreateTopics creates multiple topics
func (s *TopicServiceImpl) CreateTopics(topics []models.Topic) error {
	if err := s.topicRepo.CreateTopics(topics); err != nil {
		logrus.Errorf("Error creating topics: %v", err)
		return errors.New("failed to create topics")
	}
	return nil
}

// UpdateTopic updates a topic's name
func (s *TopicServiceImpl) UpdateTopic(topicID uint, name string) (*models.Topic, error) {
	topic := &models.Topic{}
	if err := s.topicRepo.GetByID(topicID, topic); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("topic not found")
		}
		return nil, errors.New("failed to get topic")
	}

	topic.Name = utils.SanitizeString(name)
	if err := s.topicRepo.Update(topic); err != nil {
		logrus.Errorf("Error updating topic: %v", err)
		return nil, errors.New("failed to update topic")
	}

	return topic, nil
}

// DeleteTopic deletes a topic
func (s *TopicServiceImpl) DeleteTopic(topicID uint) error {
	if err := s.topicRepo.Delete(topicID, &models.Topic{}); err != nil {
		logrus.Errorf("Error deleting topic: %v", err)
		return errors.New("failed to delete topic")
	}
	return nil
}

// DeleteTopicsByChapterID deletes all topics for a chapter
func (s *TopicServiceImpl) DeleteTopicsByChapterID(chapterID uint) error {
	if err := s.topicRepo.DeleteTopicsByChapterID(chapterID); err != nil {
		logrus.Errorf("Error deleting topics by chapter ID: %v", err)
		return errors.New("failed to delete topics")
	}
	return nil
}